File: spring-data/src/main/java/dev/cerbos/queryplan/springdata/NullAttributeRepresentation.java
/*
 * Copyright 2021-2026 Zenauth Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

package dev.cerbos.queryplan.springdata;

/**
 * How the caller represents a NULL column in the attributes it sends to {@code check()}.
 *
 * <p>The planner emits the same {@code eq(attr, null)} node either way, so the adapter has to
 * be told.
 */
public enum NullAttributeRepresentation {

    /**
     * A NULL column is sent as an explicit {@code null} attribute, so {@code IS NULL} selects
     * exactly the rows {@code check()} allows. The default.
     */
    EXPLICIT,

    /**
     * A NULL column is sent as no attribute at all, which Cerbos denies as a missing-attribute
     * error. Null comparison operands are then rejected rather than translated to
     * {@code IS NULL}.
     */
    OMITTED
}

File: exposed/src/main/kotlin/dev/cerbos/queryplan/exposed/NullAttributeRepresentation.kt
package dev.cerbos.queryplan.exposed

/**
 * How the caller represents a SQL `NULL` column in the attributes it sends to `check()`.
 *
 * The planner emits the same `eq(attr, null)` node under both conventions, so a plan cannot reveal
 * which one the caller uses and the adapter has to be told
 * (https://github.com/cerbos/query-plan-adapters/issues/302). Set the call-level default on
 * [Options], and override it per attribute on [AttributeMapping.Field] where one policy suite mixes
 * the two (https://github.com/cerbos/query-plan-adapters/issues/308).
 */
public enum class NullAttributeRepresentation {
    /**
     * A NULL column is sent as an explicit `null` attribute. `IS NULL` then selects exactly the
     * rows `check()` allows. The default.
     */
    EXPLICIT,

    /**
     * A NULL column sends no attribute at all, so CEL raises a missing-attribute error (a deny) and
     * a NULL-selecting filter would return rows the PDP denies. Null operands are refused.
     */
    OMITTED,
}
